### PR TITLE
fixed td-agent version 3 for amazon linux

### PR DESCRIPTION
**Pull request by wannesvc**
Hi,
I've experienced some issue trying to install td-agent v3 on Amazon linux. I've made a fix that seems to work, but the creation of the repo url is kinda hacky as not all versions of amazon linux have an installer available. Could you take a look? Any feedback for improvement is most welcome!
(tested on serveral versions of Amazon linux 1 and 2)

### CODE REVIEW

File: recipes/install.rb [hunk 1]
```diff
@@ -100,8 +101,19 @@
       else
         "http://packages.treasuredata.com/#{major}/redhat/$releasever/$basearch"
       end
+    when '3'
+      if platform == "amazon"
+        amazon_version = node['platform_version'] == '2' ? '2' : '1'
+        if node['td_agent']['yum']['amazon'][amazon_version].include? node['platform_version']
+          package_version = node['platform_version']
+        else
+          package_version = node['td_agent']['yum']['amazon'][amazon_version].last
```
Comment (yyuu) [inline recipes/install.rb]: Would this mean that the cookbook installs the latest built packages if the platform is something unknown....? I don't think it's appropriate behavior.

Even if you want the cookbook to behave so, it should be opt-in explicitly by declaring custom attribute, e.g. `node["td_agent"]["yum_amazon_releasever"]`. I think some change like follows would be sufficient to use dedicated repository for Amazon Linux. You'd need to declare `node["td_agent"]["yum_amazon_releasever"]` by your own if you'd like to install td-agent3 for unsupported Amazon Linux versions, though.

```rb
when '3'
  if platform == "amazon
    amazon_version = node["platform_version"]
    "http://packages.treasuredata.com/#{major}/amazon/#{amazon_version}/#{node["td_agent"]["yum_amazon_releasever"]}/$basearch"
  else
    "http://packages.treasuredata.com/#{major}/redhat/$releasever/$basearch"
  end
  ( ... )
```

cc: @repeatedly

**Comment by wannesvc** [comment]
Hi, thanks for reviewing.
the reason I have
`amazon_version = node['platform_version'] == '2' ? '2' : '1'`
Is because for amazon linux v1, this value is something like "2018.03", while for v2 this is simply '2'.
As we need to have a 'major' version to use in the download path, we need to deduct it in some kind of way.
I adapted the code so it doesn't automatically install a version that is not supported. I hope this is what you meant?

kind regards,
Wannes

File: recipes/install.rb [hunk 2]
```diff
@@ -100,8 +101,14 @@
       else
         "http://packages.treasuredata.com/#{major}/redhat/$releasever/$basearch"
       end
+    when '3'
+      if platform == "amazon"
+        amazon_version = node['platform_version'] == '2' ? '2' : '1'
```
Comment (yyuu) [inline recipes/install.rb]: I'm not sure how the `platform_version` of future releases of Amazon Linux will be, but this kind of assumption will likely be problematic in some distant future.

How about to prepare some mapping of `platform_version` to Amazon Linux's release version as a `Hash` like follows...?

```rb
platform_version = node['platform_version']
amazon_version = node['td_agent']['yum_amazon_platform_version_mapping'][platform_version] || platform_version
```

To make the above working, we'd need to have some attribute definition like follows. With this way, users of cookbook can override the default mapping value if it is somehow insufficient.

```rb
default['td_agent']['yum_amazon_platform_version_mapping'] = {
  '2017.03' => '1',
  '2017.09' => '1',
  '2018.03' => '1',
  '2017.12' => '2',
}
```

**Comment by wannesvc** [comment]
in the end, I've switched to just checking the kernel version. I think that's a more future-proof way of determining the major amzn linux version as they won't change that name and we don't have to keep a list of versions up to date.

**Comment by wannesvc** [comment]
Hi, I think I've improved the way we check for the amazon linux version now. Could you take a look? Thx

**Pull request merged**